### PR TITLE
feat: emitter supports more complicate situation

### PR DESCRIPTION
**Pull request by e-cloud**
all the changes are made to generate types for highcharts jsdoc,
while keep the generalization of emitter.

Major changes:

* move the warn functions to logger.ts
* improve some types for the code
* handle nested properties for type properties and function params
* add a script to crawl the global types from MDN, for generating
  a list of global type reference
* `_resolveTypeString` can handle more types now
* split `handleFlags` and put it to where it should be

### CODE REVIEW

**Comment by englercj** [comment]
Looks good! Thanks for all the changes. Could you add a few of the new cases you cover to some of the test files?

The idea is to turn those into automated tests at some point, but right now it is useful to use them for manual testing.

**Comment by e-cloud** [comment]
@englercj I've altered the test structure and add some cases to it.

Besides, we should add more cases in the future.

**Comment by englercj** [comment]
Thanks! This is great.

**Comment by scott-ho** [comment]
@englercj could you release a beta version? I have another project depending on the merged code.

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "tsd-jsdoc",
   "version": "2.0.0-beta.3",
   "description": "Compiles JSDoc annotated javascript into a Typescript Declaration File (.d.ts).",
-  "main": "publish.js",
+  "main": "dist/publish.js",
```
Comment (scott-ho) [inline package.json]: @englercj I may make a mistake here. Not sure whether you release the dist dir of the repo itself.

**Comment by e-cloud** [comment]
@englercj I'm afraid using git url is not possible. Since the published code has been compiled by typescript, If use want to use the latest code, we have to make a new beta release.

**Comment by englercj** [comment]
beta 4 is out.

**Pull request merged**